pkg/operator/controller/controller.go

```diff
@@ -0,0 +1,864 @@
+
+	if isUpgrade && cr.Status.Phase != vmimportv1alpha1.PhaseUpgrading {
+		logger.Info("Observed version is not target version. Begin upgrade", "Observed version ", cr.Status.ObservedVersion, "TargetVersion", r.operatorArgs.OperatorVersion)
+		MarkCrUpgradeHealingDegraded(cr, "UpgradeStarted", fmt.Sprintf("Started upgrade to version %s", r.operatorArgs.OperatorVersion))
+		if err := r.crUpdate(vmimportv1alpha1.PhaseUpgrading, cr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// MarkCrUpgradeHealingDegraded marks the passed CR as upgrading and degraded.
+func MarkCrUpgradeHealingDegraded(cr *vmimportv1alpha1.VMImportConfig, reason, message string) {
+	conditions.SetStatusCondition(&cr.Status.Conditions, conditions.Condition{
+		Type:   conditions.ConditionAvailable,
+		Status: corev1.ConditionTrue,
+	})
+	conditions.SetStatusCondition(&cr.Status.Conditions, conditions.Condition{
+		Type:   conditions.ConditionProgressing,
+		Status: corev1.ConditionTrue,
+	})
+	conditions.SetStatusCondition(&cr.Status.Conditions, conditions.Condition{
+		Type:    conditions.ConditionDegraded,
+		Status:  corev1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+func newDefaultInstance(obj runtime.Object) runtime.Object {
+	typ := reflect.ValueOf(obj).Elem().Type()
+	return reflect.New(typ).Interface().(runtime.Object)
+}
+
+func (r *ReconcileVMImportConfig) reconcileUpdate(logger logr.Logger, cr *vmimportv1alpha1.VMImportConfig) (reconcile.Result, error) {
+	if err := r.checkUpgrade(logger, cr); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	resources, err := r.getAllResources(cr)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var allErrors []error
+	for _, desiredRuntimeObj := range resources {
+		desiredMetaObj := desiredRuntimeObj.(metav1.Object)
+		currentRuntimeObj := newDefaultInstance(desiredRuntimeObj)
+
+		key := client.ObjectKey{
+			Namespace: desiredMetaObj.GetNamespace(),
+			Name:      desiredMetaObj.GetName(),
+		}
+		err = r.client.Get(context.TODO(), key, currentRuntimeObj)
+
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return reconcile.Result{}, err
+			}
+
+			setLastAppliedConfiguration(desiredMetaObj)
+			setLabel(createVersionLabel, r.operatorArgs.OperatorVersion, desiredMetaObj)
+
+			if err = controllerutil.SetControllerReference(cr, desiredMetaObj, r.scheme); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			currentRuntimeObj = desiredRuntimeObj.DeepCopyObject()
+			logger.Info("Resource creating",
+				"namespace", desiredMetaObj.GetNamespace(),
+				"name", desiredMetaObj.GetName(),
+				"type", fmt.Sprintf("%T", desiredMetaObj))
+
+			if err = r.client.Create(context.TODO(), currentRuntimeObj); err != nil {
+				logger.Error(err, "")
+				allErrors = append(allErrors, err)
+				continue
+			}
+
+			logger.Info("Resource created",
+				"namespace", desiredMetaObj.GetNamespace(),
+				"name", desiredMetaObj.GetName(),
+				"type", fmt.Sprintf("%T", desiredMetaObj))
```

Comment:
will it be possible to extract this to `createResourceForController` method ?
mainly since this method is pretty long

Reply:
Ack

Reply:
It is not trivial to extract this as method but I will try to reduce the size.